### PR TITLE
Add more in depth PBR lighting model

### PR DESCRIPTION
**Pull request by QuantumEntangledAndy**
This adds the following,

- Roughness to material
- Metallic to material
- Ambient occulsion to material
- A PBR lighting model that uses these properties

All of these properties can either be a single value or a texture

This mostly follows the math in https://learnopengl.com/PBR/Lighting

### CODE REVIEW

**Comment by QuantumEntangledAndy** [comment]
Added an example that shows the effect of roughness/metallic:
Screenshot 2022-01-04 at 14 19 37

**Comment by Lowentwickler** [comment]
I have some weird effect in `light` example with this code. Using Linux with latest libraries.
20220104_08h18m13s_grim

**Comment by QuantumEntangledAndy** [comment]
Hmm I don't have linux box that I can test this on. Does disabling any of the types of lights help with it?

**Comment by QuantumEntangledAndy** [comment]
I do bind to the dummy texture multiple times, maybe the linux backend dosen't like that

**Comment by Lowentwickler** [comment]
No, but effect is not that strong with zero intensity of ambient light, though still persist. Same issue in the new example. No problem, I will investigate it on linux.
20220104_08h22m27s_grim

**Comment by QuantumEntangledAndy** [comment]
I'll push a commit that uses unique dummy textures

**Comment by QuantumEntangledAndy** [comment]
Ok using unique dummy textures has been pushed

**Comment by Lowentwickler** [comment]
On my mac it is also not working properly.

Snímek obrazovky 2022-01-04 v 11 24 57
Snímek obrazovky 2022-01-04 v 11 25 27

**Comment by QuantumEntangledAndy** [comment]
How odd did you do any digging into it? I would usually debug this by commenting out things in the wgsl and the activating it one bit at a time.

**Comment by Lowentwickler** [comment]
I will dive into it later today. Just checked on different machines between other tasks. My colleague also reported that on Windows it looks wrong. That is good actually, we can easily reproduce it, so fix won't be that hard. 

BTW, what video adapter you have on your Mac?

**Comment by QuantumEntangledAndy** [comment]
Here's my lowly mac specs :p
Screenshot 2022-01-04 at 21 29 19

Intel HD Graphics 615 1536 MB graphics

File: dotrix_pbr/src/shaders/skeletal.wgsl [hunk 1]
```diff
@@ -89,30 +89,79 @@ fn vs_main(
...
-        albedo_color = u_material.albedo;
+        metallic = u_material.metallic.color;
     }
-    let light_color = calculate_light(in.world_position.xyz, in.normal);
 
-    return albedo_color * light_color;
+    if (u_material.ao.has_texture != 0u) {
+        metallic = textureSample(r_ao_texture, r_sampler, in.tex_uv);
+    } else {
+        metallic = u_material.ao.color;
```
Comment (Lowentwickler) [inline dotrix_pbr/src/shaders/skeletal.wgsl:156]:
```
ao = 
```

this is the issue: undefined variable

File: dotrix_pbr/src/shaders/solid.wgsl [hunk 2]
```diff
@@ -44,31 +44,79 @@ fn vs_main(
...
     } else {
-        albedo_color = u_material.albedo;
+        metallic = u_material.metallic.color;
     }
 
-    let light_color = calculate_light(in.world_position, in.normal);
+    if (u_material.ao.has_texture != 0u) {
+        metallic = textureSample(r_ao_texture, r_sampler, in.tex_uv);
+    } else {
+        metallic = u_material.ao.color;
```
Comment (Lowentwickler) [inline dotrix_pbr/src/shaders/solid.wgsl:111]:
```
ao = 
```

this is the issue: undefined variable

File: dotrix_pbr/src/shaders/light.inc.wgsl [hunk 3]
```diff
@@ -70,104 +76,210 @@ fn calculate_point(
...
+    let NdotV: f32 = max(dot(N, V), 0.0);
+    let NdotL: f32 = max(dot(N, L), 0.0);
+    let ggx2: f32  = GeometrySchlickGGX(NdotV, roughness);
+    let ggx1: f32  = GeometrySchlickGGX(NdotL, roughness);
+
+    return ggx1 * ggx2;
 }
 
+// Calulates the amount of light that refects (specular) and that which scatters (diffuse)
+fn fresnelSchlick(cosTheta: f32, F0: vec3<f32>) -> vec3<f32>
```
Comment (Lowentwickler) [inline dotrix_pbr/src/shaders/light.inc.wgsl:151]: could you please bring all new functions names to a snake_case?

**Review by Lowentwickler — CHANGES_REQUESTED** [review]

File: dotrix_pbr/src/material.rs [hunk 4]
```diff
@@ -46,10 +120,21 @@ impl Material {
 
 #[repr(C)]
 #[derive(Default, Debug, Clone, Copy)]
-struct Uniform {
-    albedo: [f32; 4],
+struct MaterialUniform {
+    color: [f32; 4],
     has_texture: u32,
-    reserve: [u32; 3],
+    reverve: [f32; 3],
```
[the remark this answers is not in the document]
Comment (QuantumEntangledAndy) [reply dotrix_pbr/src/material.rs]: Done

File: dotrix_pbr/src/material.rs [hunk 5]
```diff
@@ -12,31 +15,102 @@ pub struct Material {
...
+            .find::<Texture>(DUMMY_AO_TEXTURE)
+            .expect("System `dotrix::pbr::material::startup` must be executed");
         if self.texture.is_null() {
-            self.dummy_texture = true;
-            self.texture = assets.find(DUMMY_TEXTURE)
-                .expect("System `dotrix::pbr::material::startup` must be executed");
+            self.texture = dummy_id;
+        }
+        if self.roughness_texture.is_null() {
+            self.roughness_texture = dummy_id;
```
[the remark this answers is not in the document]
Comment (QuantumEntangledAndy) [reply dotrix_pbr/src/material.rs]: Done

File: dotrix_pbr/src/shaders/skeletal.wgsl (hunk 1 above)
Comment (QuantumEntangledAndy) [reply dotrix_pbr/src/shaders/skeletal.wgsl]: Ah that would explain the random nature, thanks for finding this

File: dotrix_pbr/src/shaders/light.inc.wgsl (hunk 3 above)
Comment (QuantumEntangledAndy) [reply dotrix_pbr/src/shaders/light.inc.wgsl]: Now in snake_case, I tired to add their full names too when possible

**Comment by QuantumEntangledAndy** [comment]
Ok this should be ready now :). One last thing to consider adding perhaps would be a normal (bump) map

File: dotrix_pbr/src/material.rs [hunk 6]
```diff
@@ -4,6 +4,9 @@ use dotrix_core::renderer::UniformBuffer;
 use dotrix_core::{Assets, Color, Id, Renderer};
 
 const DUMMY_TEXTURE: &str = "dotrix::dummy_texture";
+const DUMMY_ROUGHNESS_TEXTURE: &str = "dotrix::dummy_roughness_texture";
```
Comment (Lowentwickler) [inline dotrix_pbr/src/material.rs:7]: Maybe we should rollback to a single dummy texture now, what do you think?

Comment (QuantumEntangledAndy) [reply dotrix_pbr/src/material.rs:7]: Yeah we can try it

Comment (QuantumEntangledAndy) [reply dotrix_pbr/src/material.rs]: It's done, maybe you could have a quick check

File: dotrix_pbr/src/shaders/skeletal.wgsl [hunk 7]
```diff
@@ -89,30 +89,79 @@ fn vs_main(
...
-    albedo: vec4<f32>;
-    has_texture: u32;
+  color: vec4<f32>;
+  has_texture: u32;
+};
+struct Materials {
+    albedo: Material;
+    roughness: Material;
+    metallic: Material;
+    ao: Material;
```
Comment (Lowentwickler) [inline dotrix_pbr/src/shaders/skeletal.wgsl:99]: What is the reason behind this layout? What do you think about to optimize it like this
```
struct Material {
    color: vec4<f32>;
    has_texture: vec4<u32>; // x - albedo, y - roughness, z - metallic, w - ao or even pack them in bits
    roughness: vec4<f32>;
    metallic: vec4<f32>;
    ao: vec4<f32>;
```

Comment (QuantumEntangledAndy) [reply dotrix_pbr/src/shaders/skeletal.wgsl:99]: We could pack it like this too

```glsl
struct Material {
    color: vec4<f32>;
    has_texture: vec4<u32>; // x - albedo, y - roughness, z - metallic, w - ao or even pack them in bits
    properties: vec4<f32>; // x - albedo, y - roughness, z - metallic, w - ao
};

Comment (Lowentwickler) [reply dotrix_pbr/src/shaders/skeletal.wgsl:99]: I was afraid to suggest it lol :) but that would be the best

Comment (QuantumEntangledAndy) [reply dotrix_pbr/src/shaders/skeletal.wgsl:99]: I'll try bitwise but my bit math might need some work

File: dotrix_pbr/src/skeletal.rs [hunk 8]
```diff
@@ -17,10 +17,22 @@ pub struct Entity {
...
+    pub roughness_texture: Id<Texture>,
+    /// Roughness (Random scatter)
+    pub roughness: f32,
+    /// Metallic texture asset ID
+    pub metallic_texture: Id<Texture>,
+    /// Metallic (reflectance)
+    pub metallic: f32,
+    /// Ambient occulsion texture asset ID
+    pub ao_texture: Id<Texture>,
+    // Ambient occulsion
```
Comment (Lowentwickler) [inline dotrix_pbr/src/skeletal.rs:34]: Please, update it to triple slash for documentation: `///`

File: dotrix_pbr/src/material.rs (hunk 6 above)
Comment (Lowentwickler) [reply dotrix_pbr/src/material.rs]: Looks good and works on my linux laptop

File: dotrix_pbr/src/shaders/skeletal.wgsl (hunk 7 above)
Comment (QuantumEntangledAndy) [reply dotrix_pbr/src/shaders/skeletal.wgsl]: Ok compact materials is in

File: dotrix_pbr/src/skeletal.rs (hunk 8 above)
Comment (QuantumEntangledAndy) [reply dotrix_pbr/src/skeletal.rs]: Done

File: dotrix_pbr/src/shaders/skeletal.wgsl [hunk 9]
```diff
@@ -125,43 +122,43 @@ fn average(input: vec4<f32>) -> f32 {
...
+    if ((u_material.has_texture & 2u) == 2u) {
+        roughness = average(textureSample(r_roughness_texture, r_sampler, in.tex_uv));
     } else {
-        roughness = u_material.roughness.color;
+        roughness = u_material.roughness;
     }
 
-    if (u_material.metallic.has_texture != 0u) {
-        metallic = textureSample(r_metallic_texture, r_sampler, in.tex_uv);
+    if ((u_material.has_texture & 4u) == 2u) {
```
Comment (Lowentwickler) [inline dotrix_pbr/src/shaders/skeletal.wgsl:144]:
```
& 4u) == 4u
```

File: dotrix_pbr/src/shaders/solid.wgsl [hunk 10]
```diff
@@ -79,44 +76,44 @@ fn average(input: vec4<f32>) -> f32 {
...
+      albedo = u_material.albedo;
+  }
+
+  if ((u_material.has_texture & 2u) == 2u) {
+      roughness = average(textureSample(r_roughness_texture, r_sampler, in.tex_uv));
+  } else {
+      roughness = u_material.roughness;
+  }
+
+  if ((u_material.has_texture & 4u) == 2u) {
```
Comment (Lowentwickler) [inline dotrix_pbr/src/shaders/solid.wgsl:99]: `& 4u) == 4u`

**Review by Lowentwickler — CHANGES_REQUESTED** [review]

File: dotrix_pbr/src/shaders/skeletal.wgsl (hunk 9 above)
Comment (QuantumEntangledAndy) [reply dotrix_pbr/src/shaders/skeletal.wgsl]: Ah, thank you

**Review by Lowentwickler — APPROVED** [review]
Awesome!

**Comment by QuantumEntangledAndy** [comment]
Next thing I am thinking to tackle is the normal map. We will need to compute the face tangents for that

**Comment by Lowentwickler** [comment]
Once I've read about this technique, but never implemented it though.

**Comment by QuantumEntangledAndy** [comment]
There's a nice article for normal maps [here](https://learnopengl.com/Advanced-Lighting/Normal-Mapping) just planning on translating it to rust/wgsl

**Pull request merged**